### PR TITLE
Add maximized mode to markdown-textarea + CSS improvements

### PR DESCRIPTION
**Pull request by kimsible**
This PR responds to the issue https://github.com/Chocobozzz/PeerTube/issues/2656.

It adds some UI/UX optimizations to the markdown-textarea component :

- Maximized mode side-by-side (textarea / preview) edition 
- CSS improvements of the `flex-column` view inspired by the GitHub textarea
- Fix `max-height` to the preview in minimized mode to prevent from breaking display with long text
- Disable previews for small displays (`width < 800px`) but still enabled in maximized mode

**Side effets**

~The `body` scrollbar stays displayed in maximized mode, it just moves the content behind the maximized view~  
=> Add a class `.lock-scroll` on `#content` that will fix the main component.

~Swiching between portrait / landscape orientation on touch devices may hide the scroll for all views.~
~=> Add ResizeObserver and Rxjs Observer on `orientationchange` 
 to preserve content fixed / scrolling when switching maximized mode on a mobile display~

**Questions**

- The `flex-row` view (with the preview on the right on large screens) has not been changed and it is still weird on large screens, maybe only the `flex-column` view as below should be used ?

- It's just CSS / templating changes, these changes wouldn't increase too much bundle styles ?

**EDIT: only the `flex-column` view has been kept outside of maximized mode**

**TODO**

- [x] Test on Firefox
- [x] Test on Chrome
- [x] Fix others textareas
- [x] Handle locking body scroll
- [x] ~Fix ResizeObserver unknown by TypeScript compiler~ not required anymore since there is a maximized mode for mobile views
- [x] Test on WebKitGTK browser (Safari Linux)
- [x] Test on Edge (v80)

t-flex-maximized-fulltext

### CODE REVIEW

**Comment by rigelk** [comment]
This is all good. Bundle size shouldn't suffer 180+LOCs, especially since they are factorized in a reusable component.

However, the issue you raise about side-by-side editing on large screens is one that should be handled, potentially in another issue/PR.

**Comment by rigelk** [comment]
A question though: where did the icons come from?

**Comment by kimsible** [comment]
@rigelk 

> This is all good. Bundle size shouldn't suffer 180+LOCs, especially since they are factorized in a reusable component.

> However, the issue you raise about side-by-side editing on large screens is one that should be handled, potentially in another issue/PR.

Great  :), I am going to try to make the css part clearer.
Also, I ask if we really need the classical side-by-side editing, it is used only on 2 views : 

- Path `/videos/update/:uuid`, tab advanced-settings, field Support
- Path `/admin/config/edit-custom`, tab instance information, all markdown fields

Maybe, having the same display everywhere can unify the UI and improve UX.

My proposition would be to display everywhere the same `markdown-textarea` : 

- Small preview with tabs (Truncated, Complete previews and maximize button) below the textarea

OR

- Textarea tabbed with preview like GitHub does but it can be hard to understand for uninitiated to markdown if we don't have controls (Bold, Italic, Stroke, Links...)

In any case the maximized mode to see side-by-side textarea editor + preview.

I'm going to try to delete the other view on larger screen `flex-row` on this draft.

What to you think about this proposition ?

> A question though: where did the icons come from?

The icon is from https://icons.getbootstrap.com/

**Comment by kimsible** [comment]
Also I added a mobile view, but do we really need it ? 

Maybe I'm wrong but edition should require a minimum display width (800px / small-screen) ?

File: client/src/app/shared/forms/markdown-textarea.component.scss [hunk 1]
```diff
@@ -26,11 +50,139 @@
...
+
+        :last-child {
+          margin-right: $not-expanded-horizontal-margins;
+        }
+      }
+    }
+  }
+}
+
+@media screen and (min-width: $mobile-view) and (max-width: #{$small-view + 135px}) {
```
Comment (Chocobozzz) [inline client/src/app/shared/forms/markdown-textarea.component.scss:164]: Why `135px`? (maybe add a comment

File: client/src/app/shared/forms/markdown-textarea.component.ts [hunk 2]
```diff
@@ -80,6 +81,14 @@ export class MarkdownTextareaComponent implements ControlValueAccessor, OnInit {
     this.contentChanged.next(this.content)
   }
 
+  onMaximizeClick () {
+    if (this.isMaximized) {
```
Comment (Chocobozzz) [inline client/src/app/shared/forms/markdown-textarea.component.ts:85]: `this.isMaximized = !this.isMaximized`

File: client/src/app/shared/images/global-icon.component.ts [hunk 3]
```diff
@@ -54,7 +54,9 @@ const icons = {
   'users': require('!!raw-loader?!../../../assets/images/global/users.svg').default,
   'search': require('!!raw-loader?!../../../assets/images/global/search.svg').default,
   'refresh': require('!!raw-loader?!../../../assets/images/global/refresh.svg').default,
-  'npm': require('!!raw-loader?!../../../assets/images/global/npm.svg').default
+  'npm': require('!!raw-loader?!../../../assets/images/global/npm.svg').default,
+  'arrows-angle-expand': require('!!raw-loader?!../../../assets/images/global/arrows-angle-expand.svg').default,
```
Comment (Chocobozzz) [inline client/src/app/shared/images/global-icon.component.ts:58]: Please check if you can use icons from https://robbiepearce.com/softies/ to keep consistency with the rest of the project

Comment (rigelk) [reply client/src/app/shared/images/global-icon.component.ts:58]: Note that softies has fullscreen in/out icons, which we already use and could make good replacements.

**Comment by rigelk** [comment]
> Also I added a mobile view, but do we really need it ? Maybe I'm wrong but edition should require a minimum display width (800px / small-screen) ?

Mobile-friendliness is often asked, so please keep it that way.

File: client/src/app/shared/forms/markdown-textarea.component.scss (hunk 1 above)
Comment (kimsible) [reply client/src/app/shared/forms/markdown-textarea.component.scss]: Sorry, CSS has been fully refactored with good practices and without this trick

**Comment by kimsible** [comment]
Here is some new screenshots with the softies icons : 

## Mobile display (less than 500px)
t-mobile

## Small display (from 500px to 800px)
t-small-view

## Maximized mode for small display
t-flex-maximized-fulltext-small-display

## Medium display (from 800px)
t-flex-column-complete
t-flex-column-truncated

## Maximized mode medium display
t-flex-maximized-fulltext

**Comment by kimsible** [comment]
## Channel settings
t-channel-settings

## Video edit description
t-video-edit-description

## Video edit advanced settings
t-video-edit-advanced-settings

## Admin configuration about page
t-admin-config

**Comment by kimsible** [comment]
> > Also I added a mobile view, but do we really need it ? Maybe I'm wrong but edition should require a minimum display width (800px / small-screen) ?
> 
> Mobile-friendliness is often asked, so please keep it that way.
> 
> On the contrary, I think we can now drop IE support.

@rigelk Edge must be tested too since it uses a blink-chrome engine like Opera ?  I tested it too on a webkitGtk browser (Safari Linux) and it looks perfect.

**Comment by kimsible** [comment]
I now wonder it the "Maximized mode for small display" should also be available for mobile display, the preview would fit on a screen with a width > 320px. @Chocobozzz, @Chocobozzz   ?

**Comment by kimsible** [comment]
> I now wonder if the "Maximized mode for small display" should also be available for mobile display, the preview would fit on a screen with a width between 320px and 500px. @Chocobozzz, @Chocobozzz ?

It finally simplify the code a lot, no need observables / listeners on body resize or orientation change anymore.

To display tabs on mobile screens, the tabs are simply truncated with a larger space for active tab :

<div>


</div>

File: client/src/app/shared/forms/markdown-textarea.component.ts [hunk 4]
```diff
@@ -80,10 +75,33 @@ export class MarkdownTextareaComponent implements ControlValueAccessor, OnInit {
...
+    // Make sure the window has no scrollbars
+    if (!this.isMaximized) {
+      this.unlockBodyScroll()
+    } else {
+      this.lockBodyScroll()
+    }
+  }
+
   arePreviewsDisplayed () {
+    // Unused
```
Comment (rigelk) [inline client/src/app/shared/forms/markdown-textarea.component.ts:93]: If this is unused, we might as well remove it.

File: client/src/sass/include/_variables.scss [hunk 5]
```diff
@@ -93,6 +97,9 @@ $variables: (
 
   --inputBackgroundColor: var(--inputBackgroundColor),
   --inputPlaceholderColor: var(--inputPlaceholderColor),
+  --inputBorderColor: var(--inputBorderColor),
```
Comment (rigelk) [inline client/src/sass/include/_variables.scss:100]: We have plenty of other components which this added variable won't impact. I suggest you directly use the variable instead of the css custom variable, and rather make that change in a separate PR (though this really isn't a priority… :))

Comment (kimsible) [reply client/src/sass/include/_variables.scss]: Ok ! I put the sass variable in the component instead and removed added css custom / sass variables in `_mixins.scss` and `_variables.scss`

**Comment by Chocobozzz** [comment]
Please fix the build and we can merge :) https://travis-ci.com/github/Chocobozzz/PeerTube/jobs/323996727

**Comment by kimsible** [comment]
> Please fix the build and we can merge :) https://travis-ci.com/github/Chocobozzz/PeerTube/jobs/323996727

Thanks ! it's fixed :)
